ibis/mapd/compiler.py

```diff
@@ -0,0 +1,271 @@
+
+        return buf.getvalue()
+
+
+class MapDTableSetFormatter(compiles.TableSetFormatter):
+    """
+
+    """
+    _join_names = {
+        ops.InnerJoin: 'JOIN',
+        ops.LeftJoin: 'LEFT JOIN'
+    }
+
+    def get_result(self):
+        # Got to unravel the join stack; the nesting order could be
+        # arbitrary, so we do a depth first search and push the join tokens
+        # and predicates onto a flat list, then format them
+        op = self.expr.op()
+
+        if isinstance(op, ops.Join):
+            self._walk_join_tree(op)
+        else:
+            self.join_tables.append(self._format_table(self.expr))
+
+        buf = StringIO()
+        buf.write(self.join_tables[0])
+        for jtype, table, preds in zip(self.join_types, self.join_tables[1:],
+                                       self.join_predicates):
+            buf.write('\n')
+            buf.write(util.indent('{0} {1}'.format(jtype, table), self.indent))
+
+            if len(preds):
+                buf.write('\n')
+                fmt_preds = map(self._format_predicate, preds)
+                fmt_preds = util.indent('USING ' + ', '.join(fmt_preds),
+                                        self.indent * 2)
+                buf.write(fmt_preds)
+
+        return buf.getvalue()
+
+    def _validate_join_predicates(self, predicates):
+        for pred in predicates:
+            op = pred.op()
+            if not isinstance(op, ops.Equals):
+                raise com.TranslationError(
+                    'Non-equality join predicates are not supported'
+                )
+
+            left_on, right_on = op.args
+            if left_on.get_name() != right_on.get_name():
+                raise com.TranslationError(
+                    'Joining on different column names is not supported'
+                )
+
+    def _format_predicate(self, predicate):
+        column = predicate.op().args[0]
+        return column.get_name()
+
+    def _quote_identifier(self, name):
+        return name
+
+
+class MapDExprTranslator(compiles.ExprTranslator):
+    """
+
+    """
+    _registry = _operation_registry
+    context_class = MapDQueryContext
+
+    def name(self, translated, name, force=True):
+        return _name_expr(translated, name)
+
+    @classmethod
+    def compiles(cls, klass, f=None):
+        def decorator(f):
+            cls._registry[klass] = f
+
+        if f is None:
+            return decorator
+        else:
+            decorator(f)
+            return f
+
+
+class MapDDialect(compiles.Dialect):
+    """
+
+    """
+    translator = MapDExprTranslator
+
+
+dialect = MapDDialect
+compiles = MapDExprTranslator.compiles
+rewrites = MapDExprTranslator.rewrites
+
+
+@compiles(ops.StandardDev)
+def compile_std(translator, expr):
+    # pull out the arguments to the expression
+    args = expr.op().args
+    args_size = len(args)
+
+    x = args[0]
+    how = '' if args_size < 2 else '_%s' % args[1].upper()[:4]
```

Comment:
This is very hard to understand. Can you take a look at one of the other string generating backends that needs to deal with this and see if you can clean up your code based on that?